### PR TITLE
Fix function specifiers for templates

### PR DESCRIPTION
**Pull request by aneshlya**
- Produce error when template is used with `extern "C"`/ `extern "SYCL"`
- Fixed storage class for templates, so it's now extern by default as for regular ISPC functions
- Fixed linkage for different types of template instantiations (aligned with Clang behavior)

### CODE REVIEW

**Review by nurmukhametov — APPROVED** [review]
LGTM!

File: src/parse.yy [hunk 1]
```diff
@@ -2829,26 +2829,33 @@ lAddTemplateDeclaration(TemplateParms *templateParmList, DeclSpecs *ds, Declarat
...
-            Assert(m->errorCount > 0);
-            return;
-        }
+    }
+    // We can't support extern "C"/extern "SYCL" for templates because
+    // we need mangling information.
+    if (ds->storageClass == SC_EXTERN_C || ds->storageClass == SC_EXTERN_SYCL) {
+        Error(decl->pos, "Illegal linkage provided with function template.");
+        return;
+    }
```
Comment (dbabokin) [inline src/parse.yy:2840]: I think it make sense for clarity and readability to assert on possible `storageClass` values at this point. But if you prefer not to, it's not a problem.

File: src/module.cpp [hunk 2]
```diff
@@ -1114,10 +1114,12 @@ void Module::AddFunctionTemplateDeclaration(const TemplateParms *templateParmLis
 
     // TODO: inline / noinline check.
 
-    // TODO: vectorcall checks
-
     // ...
 
+    // We don't need to worry here about passing vectorcall/regcall specifier since we support these
```
Comment (dbabokin) [inline src/module.cpp:1119]: Doesn't this mean that we should check that these specifiers are not present and generate the error otherwise?

**Review by dbabokin — APPROVED** [review]
Look good, but check the remarks below.

File: src/module.cpp (hunk 2 above)
Comment (aneshlya) [reply src/module.cpp:1119]: This check is in parse.yy. I've added it to lit test as well.

**Pull request merged**